File: src/app/model/PowerBalenceMetric.ts
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";
import { PowerMeterServiceEvent } from "./PowerMeterServiceEvent";

export class PowerBalenceMetric extends Metric {
    private handler = this.handlePowerMeterEvent.bind(this);

    constructor(metricService: MetricService) {
        super(MetricType.POWER_BALENCE, 'Power Balence', '%', metricService, '%', 0);
    }

    startLogging(): void {
        this.metricService.getPowerMeterService().subscribeForPowerData(this.handler);
    }


    stopLogging(): void {
        this.metricService.getPowerMeterService().unsubscribeForPowerData(this.handler);
    }

    handlePowerMeterEvent(event: PowerMeterServiceEvent): void {
        if (event.powermeterData.balance) {
            this.addValue(event.powermeterData.balance, new Date(event.powermeterData.timestamp));
        }
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/PowerBalenceMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnPowerData;

public class PowerBalenceMetric extends Metric implements OnPowerData {
    public PowerBalenceMetric(IMetricService metricService) {
        super(MetricType.POWER_BALENCE, "Power Balence", "%", metricService, "%", 0);
    }

    @Override
    public void startLogging() {
        metricService.getPowerMeterService().subscribeForPowerData(this);
    }

    @Override
    public void stopLogging() {
        metricService.getPowerMeterService().unsubscribeForPowerData(this);
    }

    @Override
    public void onPowerData(PowerMeterData data) {
        if (data.balance != null && data.balance != 0) {
            addValue(data.balance, data.timestamp);
        }
    }
}
